app/controllers/users_controller.rb

```diff
@@ -148,6 +148,40 @@ def edit
     @breadcrumbs << (view_context.link_to @user.display_name, user_path(@user))
   end
 
+  action_auth_level :download_all_submissions, :student
+  def download_all_submissions
+    user = User.find(params[:id])
+    submissions = if params[:final]
+                    Submission.latest.where(course_user_datum: CourseUserDatum.where(user_id: user))
+                  else
+                    Submission.where(course_user_datum: CourseUserDatum.where(user_id: user))
+                  end
+    submissions = submissions.select do |s|
+      p = s.handin_file_path
+      is_disabled = s.course_user_datum.course.disabled
+      !p.nil? && File.exist?(p) && File.readable?(p) && !is_disabled
+    end
+    if submissions.empty?
+      flash[:error] = "There are no submissions to download."
+      redirect_to(user_path(user)) && return
+    end
+    filedata = submissions.collect do |s|
+      p = s.handin_file_path
+      course_name = s.course_user_datum.course.name
+      assignment_name = s.assessment.name
+      [p, download_filename(p, course_name, assignment_name)]
+    end
+    result = Archive.create_zip filedata
+    if result.nil?
+      flash[:error] = "There are no submissions to download."
+      return
+    end
+    send_data(result.read, # to read from stringIO object returned by create_zip
+              type: "application/zip",
+              disposition: "attachment", # tell browser to download
+              filename: "autolab_submissions.zip")
```

Comment:
nit:  you can add the date or some unique string to the filename so that if the user downloads multiple times it's not the same file name

Reply:
Done in recent commit